Basic nn classes and functionality

chrisc2244 opened the pull request:
Created basic node class and basic layer class, functionality... to be determined.

chrisc2244 committed:
Created basic Node class

chrisc2244 committed:
Created basic layer class

chrisc2244 committed:
Adding matrix class. Matrix constructor works, also includes printing correctly.
Need to add more matrix functionality and operator overloads

chrisc2244 committed:
Added convertPixelToGreyscale() and convertPixelVectorToGreyscaleVector() to ImageWrapper

chrisc2244 committed:
Added matrix functionality for double vectors (meant for normalized greyscale)

Added constructor for matrix class that takes a double vector

chrisc2244 committed:
removed delete[] to stop deleting the matrix literally as it's created.
added operator[] overload for matrixes

chrisc2244 committed:
added operator+ overload for adding matrix objects together, throws exception if matrices are different sizes

added operator() overload for access element in matrix in (row, col) form

fixed operator[] overload, now correctly accesses element in matrix at index, throws exception if accessing out of bounds

added static matrix::print() method that takes a matrix and prints it nicely into the terminal

added a few public variables (size, p_columns, p_rows) for the methods to work... this prob needs to be changed at some point

chrisc2244 committed:
fixed matrix operator+ method, works properly now 100% tested!!

added matrix operator- method, also works properly 100% tested!

chrisc2244 committed:
added more bigger if-statement check for overloaded matrix operators to avoid trying to do math on matrixes that have same size but different orientations, like trying to subtract 3x2 by 2x3 matrix

chrisc2244 committed:
added better messages on exception throws

operator() overload yet to work correctly, just committed it so I can get the code on my pc, will fix

Jon-Cavaliere commented on src/Matrix.h:
```diff
@@ -0,0 +1,59 @@
+	//access element in matrix in (row, col) form
+	double& operator() (int row, int col) const;
+
+	int size;
```
Why do we have public and private size variables of different datatypes?

Jon-Cavaliere replied:
Perhaps keep the private ones, and just use some getter functions? uint32_t size()

uint32_t getNumRows() etc...?

chrisc2244 replied:
467934ac42a33b26e2cb83b6ccab2687a6ae9c28

removed public variables and their usages, changed all usages to member variables and using getter methods when appropriate

Jon-Cavaliere commented on src/Matrix.h:
```diff
@@ -0,0 +1,59 @@
+
+private:
+
+	uint8_t m_rows;
```
uint8_t means our max matrix size is 255x255 was this intentional?

The reason I ask is because there seems to be some functionality in here related to pixel data

Jon-Cavaliere replied:
I would suggest decoupling the image stuff from this class completely and just having a pure matrix representation with functionality.

chrisc2244 replied:
467934ac42a33b26e2cb83b6ccab2687a6ae9c28

decoupled from ImageWrapper, also I assume our matrix won't get bigger than 255x255 but we can change that if we need to

chrisc2244 committed:
added operator() overload for accessing matrix in (row, col) form

Jon-Cavaliere commented on src/Matrix.cpp:
```diff
@@ -0,0 +1,181 @@
+*/
+
+
+uint8_t matrix::get_row_amount() const
```
if we already have these functions, why do we need p_columns, and p_rows?

chrisc2244 replied:
467934ac42a33b26e2cb83b6ccab2687a6ae9c28

same as above, public variables removed

Jon-Cavaliere commented on src/Matrix.cpp:
```diff
@@ -0,0 +1,181 @@
+matrix::matrix() : size(0), p_columns(0), p_rows(0), m_rows(0), m_columns(0), m_input_pixel_vector(0),
+m_size(0), m_ptr_double(nullptr) {}
+
+matrix::matrix(const uint8_t rows, const uint8_t columns, const std::vector<double> &greyscalevalues)
```
You are passing greyscale values as a const reference which is good. No copy is made here but see next comment V

Jon-Cavaliere commented on src/Matrix.cpp:
```diff
@@ -0,0 +1,181 @@
+	size = m_size;
+	p_columns = m_columns;
+	p_rows = m_rows;
+	m_greyscale_values = greyscalevalues;
```
Here you are copying everything into a new vector, that I'm really not sure belongs as a member of this class. Because we are storing all the data as an array on the heap

chrisc2244 replied:
467934ac42a33b26e2cb83b6ccab2687a6ae9c28

removed this when decoupled from the ImageWrapper class

Jon-Cavaliere commented on src/Matrix.cpp:
```diff
@@ -0,0 +1,181 @@
+			new_values.push_back(added_value);
+		}
+
+		matrix matrix_to_return(other_matrix.m_rows, other_matrix.m_columns, new_values);
```
Okay so this is interesting, you're creating a vector of doubles, then you populate it with the values to reflect the operation performed, then you pass it to the matrix constructor, which allocates the memory again on the heap, then copies all the values over, and returns a matrix object. 

This is unnecessarily complex, it would be better practice to just have an overloaded matrix constructor that takes a pointer to a heap allocated array instead of a vector. You could create that array here and just give the new matrix object the address of it. That way you would avoid allocating memory twice and copying the values over.

chrisc2244 replied:
a9a827fdbf777eb374609c41014bf22ceee10f2e

fixed, read the comment in that commit for details

Jon-Cavaliere commented on src/Matrix.cpp:
```diff
@@ -0,0 +1,181 @@
+{
+	if (other_matrix.size == this->size && other_matrix.m_columns == this->m_columns && other_matrix.m_rows == this->m_rows)
+	{
+		std::vector<double> new_values;
```
See last comment

Jon-Cavaliere commented on src/Matrix.cpp:
```diff
@@ -0,0 +1,181 @@
+{
+	if (row > 0 && col > 0 && col <= m_columns && row <= m_rows)
+	{
+		int row_of_element = row - 1;
```
I'm confused. Why are we adjusting the row and column index that we specify when calling this?

I think that as we are both programmers, it would be more intuitive for us the refer to the first element as at index (0, 0).

chrisc2244 replied:
70d1b772772855d147f5fc50ae4e26de80df28fc

changed so element 0 is at (0,0) now

Jon-Cavaliere commented on src/Matrix.h:
```diff
@@ -0,0 +1,58 @@
+
+	//constructors
+	matrix();
+
```
Maybe add a destructor where you deallocate the array stored as m_doubleptr I believe?

delete[] m_doubleptr;

chrisc2244 committed:
Removed public variables, using getter methods instead where needed, also decoupled ImageWrapper from matrix class entirely

chrisc2244 committed:
added constructor that takes a pointer to double array

operator+ and operator- now create a new heap-allocated double* array and uses that to construct the new matrix from the operation result, using the new matrix constructor that takes a pointer to double array. no longer creates a new vector and fills that out for the vector constructor.

gitignore update is random extension i just found, ignore it

chrisc2244 committed:
changed operator() so element 0 starts at (0,0)

chrisc2244 committed:
fake committ just getting files on my other pc

chrisc2244 committed:
Added operator* overload for multiplying matrices, returns their dot product

Some name reformatting (more needs to be done)

Jon-Cavaliere committed:
Updated Matrix

Added toString static function
Changed columns and rows datatype to uint16_t
Updated to reflect naming convention

Testing some features in Application->update()
